src/main/java/io/strimzi/kafka/bridge/mqtt/core/MqttKafkaMapper.java

```
@@ -0,0 +1,161 @@
+        this.rules = rules;
+        this.rules.sort(Comparator.comparing(MappingRule::getMqttTopicPatternLevels).reversed());
+        buildRegex();
+    }
+
+    /**
+     * Maps an MQTT topic to a Kafka topic. The topic is mapped according to the defined mapping rules.
+     *
+     * @param mqttTopic represents an MQTT topic.
+     * @return a valid Kafka topic.
+     * @see MappingRule
+     */
+    public String map(String mqttTopic) {
+        for (MappingRule rule : this.rules) {
+            Matcher matcher = this.patterns.get(this.rules.indexOf(rule)).matcher(mqttTopic);
+            HashMap<String, String> placeholders = new HashMap<>();
+
+            if (matcher.matches()) {
+                String mappedKafkaTopic = rule.getKafkaTopicTemplate();
+                String[] mqttTopicPatternParts = rule.getMqttTopicPattern().split("/");
+
+                // find MQTT_TOPIC_PLACEHOLDER_REGEX in the kafkaTopicTemplate.
+                Matcher placeholderMatcher = patterns.get(patterns.size() - 1).matcher(rule.getKafkaTopicTemplate());
+                while (placeholderMatcher.find()) {
+                    placeholders.put(placeholderMatcher.group(), null);
+                }
+
+                for (String placeholderKey : mqttTopicPatternParts) {
+                    if (placeholderKey.matches(MQTT_TOPIC_PLACEHOLDER_REGEX) && placeholders.containsKey(placeholderKey)) {
+                        placeholders.put(placeholderKey, matcher.group(removeBrackets(placeholderKey)));
+                    }
+                }
+
+                for (Map.Entry<String, String> entry : placeholders.entrySet()) {
+                    if (entry.getValue() != null) {
+                        mappedKafkaTopic = mappedKafkaTopic.replace(entry.getKey(), entry.getValue());
+                    } else {
+                        throw new IllegalArgumentException("The placeholder " + entry.getKey() + " was not assigned any value.");
+                    }
+                }
```

Would it make sense to scan for missing placeholder values during the initial scan of the MappingRules when we are building our Pattern list and throw the exception then?

---

@kyguy let me think about.